File: src/NuGetTrends.Web/Portal/src/app/shared/models/package-models.ts
// TODO: This can be used for Framework results as well.. so maybe rename to INuGetSearchResult?
export interface IPackageSearchResult {
  packageId: string;
  // JavaScript number can safely handle integers up to Number.MAX_SAFE_INTEGER (2^53-1 ≈ 9 quadrillion)
  // This is well beyond typical package download counts, even for packages exceeding int.MaxValue (2.1 billion)
  downloadCount: number;
  iconUrl: string;
}

export class PackageSearchResult implements IPackageSearchResult {
  packageId: string;
  downloadCount: number;
  iconUrl: string;

  constructor(packageId: string, downloadCount: number, iconUrl: string) {
    this.packageId = packageId;
    this.downloadCount = downloadCount;
    this.iconUrl = iconUrl;
  }
}

export interface IPackageDownloadHistory {
  id: string;
  downloads: Array<IDownloadStats>;
  color?: string;
}

export interface IDownloadStats {
  week: Date;
  // JavaScript number can safely handle integers up to Number.MAX_SAFE_INTEGER (2^53-1)
  count: number;
}

export enum SearchType {
  NuGetPackage = 1,
  Framework
}

export interface IPackageColor {
  id: string;
  color: string;
}

export interface ITrendingPackage {
  packageId: string;
  // JavaScript number can safely handle integers up to Number.MAX_SAFE_INTEGER (2^53-1)
  downloadCount: number;
  growthRate: number | null;
  iconUrl: string;
  gitHubUrl: string | null;
}

export class TagColor {
  code: string;
  private used: boolean;

  constructor(code: string, used: boolean = false) {
    this.code = code;
    this.used = used;
  }

  isInUse(): boolean {
    return this.used;
  }

  setUsed(): void {
    this.used = true;
  }

  setUnused(): void {
    this.used = false;
  }
}

export class SearchPeriod {
  text: string;
  value: number;

  constructor() {
    this.text = '';
    this.value = 0;
  }
}

// NuGet Trends has data starting from January 2012
const DATA_START_DATE = new Date(2012, 0, 1);

function calculateAllTimeMonths(): number {
  const now = new Date();
  const months = (now.getFullYear() - DATA_START_DATE.getFullYear()) * 12
    + (now.getMonth() - DATA_START_DATE.getMonth());
  return months;
}

const DefaultSearchPeriods: Array<SearchPeriod> = [
  {value: 3, text: '3 months'},
  {value: 6, text: '6 months'},
  {value: 12, text: '1 year'},
  {value: 24, text: '2 years'},
  {value: 60, text: '5 years'},
  {value: 120, text: '10 years'},
  {value: calculateAllTimeMonths(), text: 'All time'}
];

const InitialSearchPeriod: SearchPeriod = DefaultSearchPeriods[3];

export { DefaultSearchPeriods, InitialSearchPeriod };

File: src/NuGetTrends.Web.Client/Models/PackageModels.cs
namespace NuGetTrends.Web.Client.Models;

/// <summary>
/// Search result from the package search API.
/// </summary>
public record PackageSearchResult
{
    public required string PackageId { get; init; }
    public long DownloadCount { get; init; }
    public string IconUrl { get; init; } = "https://www.nuget.org/Content/gallery/img/default-package-icon.svg";
}

/// <summary>
/// Download history for a package.
/// </summary>
public record PackageDownloadHistory
{
    public required string Id { get; init; }
    public required IReadOnlyList<DownloadStats> Downloads { get; init; }
    public string? Color { get; set; }
}

/// <summary>
/// Weekly download statistics.
/// </summary>
public record DownloadStats
{
    public DateTime Week { get; init; }
    public long? Count { get; init; }
}

/// <summary>
/// Single data point for the ApexCharts trend line.
/// </summary>
public record DownloadDataPoint
{
    public DateTime Timestamp { get; init; }
    public decimal Count { get; init; }
}

/// <summary>
/// Trending package from the API.
/// </summary>
public record TrendingPackage
{
    public required string PackageId { get; init; }
    public long DownloadCount { get; init; }
    public double? GrowthRate { get; init; }
    public string IconUrl { get; init; } = "https://www.nuget.org/Content/gallery/img/default-package-icon.svg";
    public string? GitHubUrl { get; init; }
}

/// <summary>
/// Package with its assigned chart color.
/// </summary>
public record PackageColor
{
    public required string Id { get; init; }
    public required string Color { get; init; }
}

/// <summary>
/// Available search periods.
/// </summary>
public record SearchPeriod
{
    public required string Text { get; init; }
    public required int Value { get; init; }
}

/// <summary>
/// Helper class for managing search periods.
/// </summary>
public static class SearchPeriods
{
    // NuGet Trends has data starting from January 2012
    private static readonly DateTime DataStartDate = new(2012, 1, 1);

    private static int CalculateAllTimeMonths()
    {
        var now = DateTime.UtcNow;
        return (now.Year - DataStartDate.Year) * 12 + (now.Month - DataStartDate.Month);
    }

    public static readonly IReadOnlyList<SearchPeriod> Default =
    [
        new() { Value = 3, Text = "3 months" },
        new() { Value = 6, Text = "6 months" },
        new() { Value = 12, Text = "1 year" },
        new() { Value = 24, Text = "2 years" },
        new() { Value = 60, Text = "5 years" },
        new() { Value = 120, Text = "10 years" },
        new() { Value = CalculateAllTimeMonths(), Text = "All time" }
    ];

    public static readonly SearchPeriod Initial = Default[3]; // 2 years
}

/// <summary>
/// Theme preference options.
/// </summary>
public enum ThemePreference
{
    System,
    Light,
    Dark
}
